src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,259 @@
+/-
+Copyright (c) 2020 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kyle Miller.
+-/
+import combinatorics.simple_graph.basic
+import algebra.big_operators.basic
+import data.nat.parity
+import data.zmod.basic
+import tactic.omega
+import tactic.fin_cases
+/-!
+# Degree-sum formula and handshaking lemma
+
+The degree-sum formula is that the sum of the degrees of a finite
+graph is equal to twice the number of edges.  The handshaking lemma is
+a corollary, which is that the number of odd-degree vertices is even.
+
+## Main definitions
+
+- A `dart` is a directed edge, consisting of a vertex and an edge incident to it.
+- `simple_graph.sum_degrees_eq_twice_card_edges` is the degree-sum formula.
+- `simple_graph.card_odd_degree_vertices_is_even` is the handshaking lemma.
+- `simple_graph.card_odd_degree_vertices_ne_is_odd` is that the number of odd-degree
+  vertices different from a given odd-degree vertex is odd.
+- `simple_graph.exists_ne_odd_degree_if_exists_odd` is that the existence of an
+  odd-degree vertex implies the existence of another one.
+
+## Implementation notes
+
+We give a combinatorial proof by using the fact that the map from
+darts to vertices has fibers whose cardinalities are the degrees and
+that the map from darts to edges is 2-to-1.
+
+## Tags
+
+simple graphs, sums
+-/
+open finset
+
+open_locale big_operators
+
+namespace simple_graph
+universes u v
+variables {V : Type u} (G : simple_graph V)
+
+/-- A `dart` is a directed edge, consisting of a vertex and an edge incident to it -/
+@[reducible]
+def dart := Σ (v : V), G.incidence_set v
+
+/-- Gets the vertex at the start of the dart. -/
+def dart_vert : G.dart → V := sigma.fst
+
+/-- Gets the edge that the dart runs through. -/
+def dart_edge : G.dart → sym2 V := λ d, d.snd
```

Comment:
Does `sigma.snd` not work here?

Reply:
This somehow makes Lean automatically coerce a term of the subtype version of `G.incidence_set v` back to a `sym2 V`.

Maybe this would be clearer?
```lean
def dart_edge : G.dart → sym2 V := λ d, (d.snd : sym2 V)
```